### DISCUSSION TITLE
Django db connection inconsistent between tests and view

### DISCUSSION THREAD

**Discussion by Subaku** [fastapi/fastapi#8938] [Questions]
Hello,

I just found a workaround to my problem but wanted to see if anyone has had this issue and might have suggestions. I'm using Django purely for its ORM feature with my Postgres DB but FastAPI for everything else. Note I'm not doing anything async. 

Take a simple pytest using django-pytest:
```python
import pytest

from my_django_models import Asset

@pytest.mark.django_db
class TestAssets:
...

    def test_response(self):
        assets = AssetFactory.create_batch(2)
        print(Asset.objects.all())
        resp = self.client.get('/assets')
```
 
The view:
```python
from fastapi import APIRouter

from my_django_models import Asset

router = APIRouter()

@router.get('/assets')
def assets():
    print(Asset.objects.all())
    ...
```

The output:
```
[<Asset: Asset object (1)>, <Asset: Asset object (2)>]
<QuerySet []>
```

So the view has access to the db but it's not actually getting my data for some reason as if it has a bad or incorrect connection.
Now if I inherit from Django's `TransactionTestCase` class and continue using Starlette's TestClient it works... But I don't understand why. So for now that's what I'm doing but wanted to maybe get some clue into why this is the case.

As another note I have made sure I'm pointing at the right settings module and made sure `django.setup()` was called though I think django-pytest does this for me.

Appreciate any info.

**Comment by adambudziak** [comment]
I currently have the same problem. Interestingly, when I debug the app I can see that the connection is still fine when FastAPI validates POST request data, i.e. a query `ModelName.objects.all()` test inside
* test definition ✅ 
* Pydantic model validator method ✅ 
* a view itself ❌  

However, I performed a little experiment and created a dependency in my view, something like this: 
```python
def get_query():
    return ModelName.objects.all()
```

and if I pass it as a dependency for the view, then it again does not work. ❌ 

Also, I use `pytest` without any classes (i.e. plain `test_` functions)

**Comment by adambudziak** [comment]
It seems that the problem can also be mitigated by using `pytest.mark.django_db(transaction=True)` instead of plain `pytest.mark.django_db`

**Comment by penoux** [comment]
I have the same issue, while migrating all my Django views to FastAPI. The view is working OK at runtime, but I encounter database access issues between pytest and a FastAPI dependency.

```python
test_func.py

"""
pytest fixtures : 
  fastapp (FastAPI initialized app with routers)
  test_client (FastAPI TestClient instance)
  auth (User in DB + API credentials)
"""
def test_func(fastapp, test_client, auth):
    user = User.objects.get(username=auth.username)  # works
    user.check_password(auth.password) # works
    authenticate(username=auth.username, password=auth.password) # works

    response = test_client.get('/', auth=auth) # fails (see comments below)
    ...


router.py

@router.get("/")
def get_foo(user: User = Depends(get_current_user):
    ...


dependencies.py

basic = HTTPBasic()

def get_current_user(credentials: HTTPBasicCredentials = Depends(basic)) -> User:
    # fails (see comments below)
    user = authenticate(username=credentials.username, password=credentials.password)
    ...

```

Test cases :

1. **sqlite in-memory database** : access to ORM (User) ok in the test, fails with `no such table: auth_user` in the FastAPI dependency
2. **sqlite in-memory database + `@pytest.mark.django_db(transaction=True)` annotation on test** : access to ORM (User) ok in the test, fails with `no such table: auth_user` in the FastAPI dependency
2. **sqlite file database** : access to ORM (User) ok in the test, fails with `database table is locked: auth_user` in the FastAPI dependency
3. **sqlite file database + `@pytest.mark.django_db()` annotation on test** : access to ORM (User) ok in the test, fails with `database table is locked: auth_user` in the FastAPI dependency
4. **sqlite file database + `@pytest.mark.django_db(transaction=True)` annotation on test** : access to ORM (User) ok in the test, also OK in the FastAPI dependency

I expected the simplest case (sqlite in-memory database) to work, as it was the case with Django test client. Is there a threading issue there ?

**Comment by bameda** [comment]
I have the same problem with postgresql. Tests only work using `transactional_db` fixture or `pytest.mark.django_db(transaction=True)` or extending the class `django.test.TransactionTestCase` with my test functions.

It seems that the db is empty in "view" functions even if you create items with the ORM before the call to `client.get` function.
